File: src/main/java/com/github/nigelzor/hogs/WalkMove.java
package com.github.nigelzor.hogs;

public class WalkMove extends Move {
	private final Direction direction;
	private final Player player;
	private final Position from;
	private final Position to;

	public WalkMove(Player player, Direction direction, Position from, Position to) {
		this.player = player;
		this.direction = direction;
		this.from = from;
		this.to = to;
	}

	@Override
	public Board apply(Board input) {
		from.getPlayers().remove(player);
		to.getPlayers().add(player);
		return input;
	}
}

File: src/main/kotlin/com/github/nigelzor/hogs/WalkMove.kt
package com.github.nigelzor.hogs

import kotlin.test.assertFalse
import kotlin.test.assertTrue

public data class WalkMove(val from: (Board) -> Position, val to: (Board) -> Position): Move {
	public override fun apply(board: Board) {
		val player = board.currentPlayer

		assert(from(board).players.contains(player), "player not at source position")
		assert(!to(board).players.contains(player), "player already at destination position")

		from(board).players.remove(player)
		to(board).players.add(player)
	}
}
